lib/utilities/firebase_service.dart

```diff
@@ -108,108 +130,274 @@ class FirebaseService {
+      if (userDoc.exists) {
+        final data = userDoc.data();
+        if (data != null && data['name'] != null) {
+          userNames.add(data['name']);
+        } else {
+          userNames.add("Unknown"); // Fallback if the name is not found
+        }
+      }
+    }
+    return userNames;
+  } catch (e) {
+    print('Error fetching user names: $e');
+    return [];
+  }
+}
+  /// Uploads a profile picture to Firebase Storage and updates the Firestore user document
+Future<void> uploadProfilePicture(String userId, String filePath) async {
+  try {
+    final File imageFile = File(filePath);
+
+    // Reference to the storage location
+    final storageRef = _storage.ref().child('profile_pictures/$userId.jpg');
+
+    // Upload the file to Firebase Storage
+    await storageRef.putFile(imageFile);
+
+    // Get the download URL of the uploaded file
+    final String downloadUrl = await storageRef.getDownloadURL();
+
+    // Update the user document in Firestore with the download URL
+    await _firestore.collection('users').doc(userId).update({
+      'profilePicture': downloadUrl,
+    });
+
+    print('Profile picture uploaded successfully: $downloadUrl');
+  } catch (e) {
+    print('Error uploading profile picture: $e');
+    throw e;
+  }
+}
+
+
+  /// Deletes the profile picture from Firebase Storage and removes it from Firestore
+  Future<void> deleteProfilePicture(String userId) async {
+    try {
+      // Reference to the storage location
+      final storageRef = _storage.ref().child('profile_pictures/$userId.jpg');
+
+      // Delete the image from Firebase Storage
+      await storageRef.delete();
+
+      // Remove the profilePicture field from the Firestore user document
+      await _firestore.collection('users').doc(userId).update({
+        'profilePicture': FieldValue.delete(),
+      });
+
+      print('Profile picture deleted successfully');
+    } catch (e) {
+      print('Error deleting profile picture: $e');
+    }
+  }
+  Future<DocumentSnapshot<Map<String, dynamic>>> _getUserDocument(String userId) {
+  return _firestore.collection('users').doc(userId).get();
+}
+
+  /// Fetches the current user's profile picture URL
+  Future<String?> getProfilePicture(String userId) async {
+  try {
+    final userDoc = await _getUserDocument(userId);
+    if (userDoc.exists) {
+      final data = userDoc.data(); // Safely retrieve data
+      if (data != null && data['profilePicture'] != null) {
+        return data['profilePicture'] as String; // Return profile picture URL
+      }
+    }
+    return null; // Return null if no profile picture is found
+  } catch (e) {
+    print('Error fetching profile picture: $e');
+    return null; // Return null in case of an error
+  }
+}
+// Update user details
+  Future<void> updateUserProfile(String userId, String name, String email) async {
+    await _firestore.collection('users').doc(userId).update({
+      'name': name,
+      'email': email,
+    });
+  }
```

Comment:
You're only updating the email address on the user collection at the database, but the email in firebase authentication is still unchanged. So the user will have to still use the previous email to login. I know the professor will question this.

Reply:
We can remove the option if it's too long. 

But I would create another popup for "Change email" that has New Email and Confirm Password and do something like this:

```dart
import 'package:firebase_auth/firebase_auth.dart';

Future<void> changeUserEmail(String newEmail, String currentPassword) async {
  final user = FirebaseAuth.instance.currentUser;

  if (user == null) {
    throw FirebaseAuthException(
      code: 'user-not-logged-in',
      message: 'No user is currently signed in.',
    );
  }

  try {
    // Reauthenticate the user
    final credential = EmailAuthProvider.credential(
      email: user.email!,
      password: currentPassword,
    );

    await user.reauthenticateWithCredential(credential);

    // Update the email
    await user.updateEmail(newEmail);

    // Optionally, send a verification email to the new email
    await user.sendEmailVerification();

    print("Email updated successfully.");
  } on FirebaseAuthException catch (e) {
    print("Failed to update email: ${e.message}");
  } catch (e) {
    print("An unknown error occurred: $e");
  }
}
```